tsdb/head_append.go

```diff
@@ -300,15 +306,25 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
 	}
 
 	s.Lock()
-	if err := s.appendable(t, v); err != nil {
-		s.Unlock()
+	// TODO: if we definitely know at this point that the sample is ooo, then optimise
+	// to skip that sample from the WAL and write only in the WBL.
+	_, delta, err := s.appendable(t, v, a.headMaxt, a.minValidTime, a.oooTimeWindow)
+	if err == nil {
+		s.pendingCommit = true
+	}
+	s.Unlock()
+	if delta > 0 {
+		a.head.metrics.oooHistogram.Observe(float64(delta) / 1000)
```

Comment:
We call appendable twice for a sample - once inside Append() and once during Commit() (because the state of the sample can change between them). We will end up observing the delta twice if we do it inside appendable, so I think that is not an option right now :/
Also, this histogram assumes that the timestamp is milliseconds which the TSDB should not do. I think we should not have /1000 here and let the user do the timestamp math, WDYT @jesusvazquez?

Reply:
Sound good to me :+1: